### PR TITLE
Fix #4156 - Support nested subfolders in approved Measure directories

### LINKED ISSUE #4156: Recursive copy of the resources/ subfolder of a measure
Enhancement Request
--------------

User on unmethours here https://unmethours.com/question/49382/os-measure-subfolder/?comment=49406#post-id-49406 indicated he has many support files (eg: CSVs). Files need to be placed in the `resources/` subfolder so they are copied when cloning the measure. But it seems that copy is non recursive, so any subfolder under `resources/` (eg: `resources/data/`) is NOT copied. User would like a way to organize them better.

## Detailed Description

https://github.com/NREL/OpenStudio/blob/89408c89dc7e9f9dbf7ed258fa919ea6c2e6ce1f/src/utilities/bcl/BCLMeasure.cpp#L1106-L1135

## Possible Implementation

Make the copy recursive.

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

Fix #4156 - Support nested subfolders in approved Measure directories

+ (green) is what is now supported in addition to what was there. ! (orange) is what is still disallowed

```diff
 .
 ├── ./docs
+│   ├── ./docs/docs.rb
 │   ├── ./docs/.gitkeep
 │   └── ./docs/subfolder
+│       └── ./docs/subfolder/subfolder_file.txt
 ├── ./.git
!│   └── ./.git/index
 ├── ./.hidden_folder
!│   └── ./.hidden_folder/file.txt
 ├── ./LICENSE.md
 ├── ./measure.rb
 ├── ./measure.xml
+├── ./README.md
 ├── ./README.md.erb
 ├── ./resources
+│   ├── ./resources/resources.rb
 │   └── ./resources/subfolder
+│       └── ./resources/subfolder/subfolder_file.txt
!├── ./root_file.txt
 ├── ./subfolder
!│   └── ./subfolder/subfolder.txt
 └── ./tests
     ├── ./tests/example_model.osm
     ├── ./tests/output
!    │   └── ./tests/output/output.txt
     ├── ./tests/subfolder
+    │   └── ./tests/subfolder/subfolder_file.txt
     ├── ./tests/test_recursive_measure_test.rb
+    └── ./tests/tests.rb
```

## TODO

* [ ] TODO: do we keep ignoring the docs/ directory when we clone the measure?
* [ ] TODO: should `BCLMeasure(const openstudio::path&)` call `checkForUpdatesFiles`? Should `BCLMeasure::clone(const openstudio::path&)` call `checkForUpdatesFiles()` first so we have a clean BCLXML to use for copying the files?

### Pull Request Author

 - [x] BCL API Changes / Additions
 - [x] BCL API methods are tested (in `src/utilities/test`)
 - [x] All new and existing tests passes
 - [x] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [x] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [x] Code Style, strip trailing whitespace, etc.
 - [x] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [x] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/utilities/bcl/test/BCLMeasure_GTest.cpp [hunk 1]
```diff
@@ -218,3 +228,487 @@ TEST_F(BCLFixture, PatApplicationMeasures)
...
+    {fs::path("non_approved_subfolder/hello.rb"), false},
+    {fs::path("non_approved_subfolder/subfolder/hello.rb"), false},
+    {fs::path("non_approved_subfolder/output/hello.rb"), false},
+  };
+
+  for (const auto& [relativeFilePath, expectedAllowed] : tests) {
+    openstudio::path absoluteFilePath = measureDir / relativeFilePath;
+    EXPECT_EQ(expectedAllowed, BCLMeasure::isApprovedFile(absoluteFilePath, measureDir));
+  }
+}
```
Comment (jmarrec) [inline src/utilities/bcl/test/BCLMeasure_GTest.cpp:327]: Test isApprovedFile

**Comment by tijcolem** [comment]
@jmarrec  This looks good but I see one failing unit test 

```
The following tests FAILED:
	651 - BCLFixture.BCLMeasure (Failed)
```

I built this on my local mac as well and confirm that it is failing.  On thing that stands is the number of files in an equality test. 6 vs 5. 

```
/Users/tcoleman/git/OpenStudio/src/utilities/bcl/test/BCLMeasure_GTest.cpp:91: Failure
652: Expected equality of these values:
652:   6u
652:     Which is: 6
652:   measure2->files().size()
652:     Which is: 5
```

Would you mind looking at this failure?  Otherwise I think this good to go.

File: src/utilities/bcl/BCLFileReference.cpp [hunk 2]
```diff
@@ -35,7 +35,9 @@
 
 namespace openstudio {
 
-BCLFileReference::BCLFileReference(const openstudio::path& path, const bool setMembers) : m_path(openstudio::filesystem::system_complete(path)) {
+BCLFileReference::BCLFileReference(const openstudio::path& measureRootDir, const openstudio::path& relativePath, const bool setMembers)
+  : m_measureRootDir(openstudio::filesystem::system_complete(measureRootDir)),
+    m_path(openstudio::filesystem::system_complete(measureRootDir / relativePath)) {
```
[the remark this answers is not in the document]
Comment (kbenne) [reply src/utilities/bcl/BCLFileReference.cpp:40]: This was discussed on Slack, and I agree with @tijcolem that we can probably be a little more tolerant of occasional API changes in BCL, compared to the Model API. I'd recommend merge.

File: src/utilities/bcl/BCLMeasure.cpp [hunk 3]
```diff
@@ -46,52 +47,139 @@
...
+  openstudio::path parentPath = absoluteFilePath.parent_path();
 
-  for (const auto& path : openstudio::filesystem::directory_files(source)) {
-    const auto srcItemPath = source / path;
-    const auto dstItemPath = destination / path;
-
-    if (m_bclXML.hasFile(srcItemPath) || (xmlPath == openstudio::filesystem::system_complete(srcItemPath))) {
-      if (!openstudio::filesystem::copy_file_no_throw(srcItemPath, dstItemPath)) {
-        return false;
+  if (parentPath == measureDir) {
```
Comment (jmarrec) [inline src/utilities/bcl/BCLMeasure.cpp:120]: This is where the unit test is failing @tijcolem .

One again I got bit by the fact that `toPath("/path/to/dir/") != toPath("/path/to/dir")` in boost::filesystem::path...

File: src/utilities/bcl/BCLMeasure.cpp [hunk 4]
```diff
@@ -1104,31 +992,89 @@ bool BCLMeasure::save() const {
...
+  // Currently the BCLMeasure(const path&) ctor **does NOT** call checkUpdatesFiles()
+  //
+  // We could call it here by that would require a const_cast and it would increment the measure.xml versionId of the original measure,
+  // and update the list of files, which isn't a big deal if the user doesn't then save it too? Even if the user did it, wouldn't we be helping out anyways?
+
+  // If this is really not wanted, we can go back to scanning the files on disk, with a similar logic as in checkForUpdatesFiles:
+  // loop on all files on disk, and if isApproved is true, then copy them
+
+  // Copy the measure.xml
+  openstudio::filesystem::copy_file_no_throw(m_directory / openstudio::path{"measure.xml"}, newDir / openstudio::path{"measure.xml"});
```
Comment (jmarrec) [inline src/utilities/bcl/BCLMeasure.cpp:1029]: *  [ ] TODO: we need to decide and keep one of these path @kbenne @tijcolem

File: src/utilities/bcl/BCLMeasure.cpp (hunk 3 above)
Comment (jmarrec) [reply src/utilities/bcl/BCLMeasure.cpp:120]: (Path handling is absolutely infuriating in C++, whether that's boost::filesystem or std::filesystem)

Comment (jmarrec) [reply src/utilities/bcl/BCLMeasure.cpp:120]: https://github.com/NREL/OpenStudio/pull/4417/commits/5eff870ab4fb71152a81c2e386773aa4570f17bf should have fixed this.

**Review by tijcolem — APPROVED** [review]

**Pull request merged**